File: dotnet/src/Sync.cs
// `skills experimental_sync` — crawl node_modules for skills (port of sync.ts).

using System.Text.Json.Nodes;

namespace Skills;

internal sealed class SyncOptions
{
    public List<string>? Agent { get; set; }
    public bool Yes { get; set; }
    public bool Force { get; set; }
}

internal static class SyncCommand
{
    public static string ShortenPath(string full, string cwd)
    {
        var home = Sys.HomeDir();
        if (full == home || full.StartsWith(home + NodePath.Sep, StringComparison.Ordinal)) return "~" + full[home.Length..];
        if (full == cwd || full.StartsWith(cwd + NodePath.Sep, StringComparison.Ordinal)) return "." + full[cwd.Length..];
        return full;
    }

    private static List<string>? ListNames(string dir)
    {
        try
        {
            return new DirectoryInfo(dir).EnumerateFileSystemInfos().Select(i => i.Name).ToList();
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }

    /// Skills in node_modules packages (top-level and scoped), tagged with the
    /// package they came from.
    private static List<(Skill Skill, string Package)> DiscoverNodeModuleSkills(string cwd)
    {
        var nm = NodePath.Join(cwd, "node_modules");
        var output = new List<(Skill, string)>();
        if (ListNames(nm) is not { } top) return output;

        void Process(string pkgDir, string pkg)
        {
            var rootMd = NodePath.Join(pkgDir, "SKILL.md");
            if (Fs.Exists(rootMd) && SkillDiscovery.ParseSkillMd(rootMd, false) is { } root)
            {
                output.Add((root, pkg));
                return;
            }
            foreach (var search in new[] { pkgDir, NodePath.Join(pkgDir, "skills"), NodePath.Join(pkgDir, ".agents", "skills") })
            {
                if (ListNames(search) is not { } names) continue;
                foreach (var name in names)
                {
                    var skillDir = NodePath.Join(search, name);
                    if (!Fs.IsDir(skillDir)) continue;
                    var md = NodePath.Join(skillDir, "SKILL.md");
                    if (!Fs.Exists(md)) continue;
                    if (SkillDiscovery.ParseSkillMd(md, false) is { } s) output.Add((s, pkg));
                }
            }
        }

        foreach (var name in top)
        {
            if (name.StartsWith('.')) continue;
            var full = NodePath.Join(nm, name);
            if (!Fs.IsDir(full)) continue;
            if (name.StartsWith('@'))
            {
                if (ListNames(full) is not { } scoped) continue;
                foreach (var sn in scoped)
                {
                    var sp = NodePath.Join(full, sn);
                    if (Fs.IsDir(sp)) Process(sp, $"{name}/{sn}");
                }
            }
            else
            {
                Process(full, name);
            }
        }
        return output;
    }

    private static string? TryHash(string path)
    {
        try
        {
            return LocalLock.ComputeSkillFolderHash(path);
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException)
        {
            return null;
        }
    }

    private sealed record Result(string Skill, string Agent, bool Success, string? Canonical, string? Error);

    private static List<T> OrCancel<T>(List<T>? v)
    {
        if (v != null) return v;
        Ui.Cancel("Sync cancelled");
        Sys.Exit(0);
        return null!;
    }

    public static void Run(SyncOptions options)
    {
        Installer.ResetPopulated();
        var cwd = Sys.Cwd();
        var agentResult = DetectAgent.Detect();
        if (agentResult.IsAgent)
        {
            options.Yes = true;
            if (options.Agent is not { Count: > 0 } && DetectAgent.GetAgentType(agentResult.Name) is { } mapped)
            {
                var list = new List<string> { mapped };
                foreach (var ua in Agents.GetUniversalAgents())
                    if (!list.Contains(ua)) list.Add(ua);
                options.Agent = list;
            }
        }

        Sys.OutLine();
        if (!agentResult.IsAgent) Ui.Intro(Pc.BgCyan(Pc.Black(" skills experimental_sync ")));
        else Ui.Log.Info($"{Pc.BgCyan(Pc.Black(Pc.Bold($" {agentResult.Name} ")))} Agent detected — installing non-interactively");

        var spinner = new Ui.Spinner();
        spinner.Start("Scanning node_modules for skills…");
        var discovered = DiscoverNodeModuleSkills(cwd);
        if (discovered.Count == 0)
        {
            spinner.Stop(Pc.Yellow("No skills found"));
            Ui.Outro(Pc.Dim("No SKILL.md files found in node_modules."));
            return;
        }
        spinner.Stop($"Found {Pc.Green(discovered.Count.ToString())} skill{(discovered.Count > 1 ? "s" : "")} in node_modules");

        foreach (var (s, pkg) in discovered)
        {
            Ui.Log.Info($"{Pc.Cyan(s.Name)} {Pc.Dim($"from {pkg}")}");
            if (s.Description.Length > 0) Ui.Log.Message(Pc.Dim($"  {s.Description}"));
        }

        var localLock = LocalLock.Read(cwd);
        var toInstall = new List<(Skill Skill, string Package)>();
        var upToDate = new List<string>();
        if (options.Force)
        {
            toInstall.AddRange(discovered);
            Ui.Log.Info(Pc.Dim("Force mode: reinstalling all skills"));
        }
        else
        {
            foreach (var item in discovered)
            {
                if (localLock.Skills[item.Skill.Name] is { } existing && TryHash(item.Skill.Path) is { } h && h == Json.Str(existing, "computedHash"))
                {
                    upToDate.Add(item.Skill.Name);
                    continue;
                }
                toInstall.Add(item);
            }
            if (upToDate.Count > 0) Ui.Log.Info(Pc.Dim($"{upToDate.Count} skill{(upToDate.Count != 1 ? "s" : "")} already up to date"));
            if (toInstall.Count == 0)
            {
                Sys.OutLine();
                Ui.Outro(Pc.Green("All skills are up to date."));
                return;
            }
        }

        Ui.Log.Info($"{toInstall.Count} skill{(toInstall.Count != 1 ? "s" : "")} to install/update");

        var universal = Agents.GetUniversalAgents();
        var visibleUniversal = Agents.GetVisibleUniversalAgents();
        LockedSection<string> Locked() => new()
        {
            Title = "Universal (.agents/skills)",
            Items = visibleUniversal.Select(a => new SearchItem<string>(a, a, Agents.Get(a).DisplayName)).ToList(),
            HiddenCount = universal.Count - visibleUniversal.Count,
        };
        SearchItem<string> Item(string a) => new(a, a, Agents.Get(a).DisplayName) { Hint = Agents.Get(a).SkillsDir };

        List<string> targetAgents;
        if (options.Agent?.Contains("*") == true)
        {
            targetAgents = Agents.AllNames();
            Ui.Log.Info($"Installing to all {targetAgents.Count} agents");
        }
        else if (options.Agent is { Count: > 0 } list)
        {
            var invalid = list.Where(a => Agents.Find(a) == null).ToList();
            if (invalid.Count > 0)
            {
                Ui.Log.Error($"Invalid agents: {string.Join(", ", invalid)}");
                Ui.Log.Info($"Valid agents: {string.Join(", ", Agents.AllNames())}");
                Sys.Exit(1);
            }
            targetAgents = list;
        }
        else
        {
            spinner.Start("Loading agents…");
            var installed = Agents.DetectInstalledAgents();
            spinner.Stop($"{Agents.List.Count} agents");
            if (installed.Count == 0)
            {
                if (options.Yes)
                {
                    Ui.Log.Info("Installing to universal agents");
                    targetAgents = universal;
                }
                else
                {
                    targetAgents = OrCancel(SearchMultiselect.Run(new SearchMultiselectOptions<string>
                    {
                        Message = "Which agents do you want to install to?",
                        Items = Agents.GetNonUniversalAgents().Select(Item).ToList(),
                        Locked = Locked(),
                    }));
                }
            }
            else if (installed.Count == 1 || options.Yes)
            {
                targetAgents = [.. installed];
                foreach (var ua in universal)
                    if (!targetAgents.Contains(ua)) targetAgents.Add(ua);
            }
            else
            {
                var others = Agents.GetNonUniversalAgents().Where(installed.Contains).ToList();
                var initial = installed.Where(a => !universal.Contains(a)).Select(a => others.IndexOf(a)).Where(i => i >= 0).ToList();
                targetAgents = OrCancel(SearchMultiselect.Run(new SearchMultiselectOptions<string>
                {
                    Message = "Which agents do you want to install to?",
                    Items = others.Select(Item).ToList(),
                    InitialSelected = initial,
                    Locked = Locked(),
                }));
            }
        }

        var summary = new List<string>();
        foreach (var (s, pkg) in toInstall)
        {
            string canonical;
            try
            {
                canonical = Installer.GetCanonicalPath(s.Name, false);
            }
            catch (InvalidOperationException)
            {
                canonical = "";
            }
            summary.Add($"{Pc.Cyan(s.Name)} {Pc.Dim($"← {pkg}")}");
            summary.Add($"  {Pc.Dim(ShortenPath(canonical, cwd))}");
        }
        Sys.OutLine();
        Ui.Note(string.Join("\n", summary), "Sync Summary");

        if (!options.Yes && Ui.Confirm("Proceed with sync?") != true)
        {
            Ui.Cancel("Sync cancelled");
            Sys.Exit(0);
        }

        spinner.Start("Syncing skills…");
        var results = new List<Result>();
        foreach (var (s, _) in toInstall)
        {
            foreach (var at in targetAgents)
            {
                var r = Installer.InstallSkillForAgent(s, at, new InstallOptions { Global = false, Cwd = cwd, Mode = InstallMode.Symlink });
                results.Add(new Result(s.Name, Agents.Get(at).DisplayName, r.Success, r.CanonicalPath, r.Error));
            }
        }
        spinner.Stop("Sync complete");

        var successfulNames = results.Where(r => r.Success).Select(r => r.Skill).Distinct().ToList();

        foreach (var (s, pkg) in toInstall)
        {
            if (!successfulNames.Contains(s.Name) || TryHash(s.Path) is not { } h) continue;
            try
            {
                LocalLock.AddSkill(s.Name, new JsonObject { ["source"] = pkg, ["sourceType"] = "node_modules", ["computedHash"] = h }, cwd);
            }
            catch (Exception e) when (e is IOException or UnauthorizedAccessException)
            {
                // ignored, as in the TS implementation's best-effort lock update
            }
        }

        Sys.OutLine();
        if (successfulNames.Count > 0)
        {
            var lines = new List<string>();
            foreach (var name in successfulNames)
            {
                var first = results.First(r => r.Success && r.Skill == name);
                var pkg = toInstall.FirstOrDefault(t => t.Skill.Name == name).Package ?? "";
                lines.Add($"{Pc.Green("✓")} {name} {Pc.Dim($"← {pkg}")}");
                if (first.Canonical != null) lines.Add($"  {Pc.Dim(ShortenPath(first.Canonical, cwd))}");
            }
            var n = successfulNames.Count;
            Ui.Note(string.Join("\n", lines), Pc.Green($"Synced {n} skill{(n != 1 ? "s" : "")}"));
        }
        var failed = results.Where(r => !r.Success).ToList();
        if (failed.Count > 0)
        {
            Sys.OutLine();
            Ui.Log.Error(Pc.Red($"Failed to install {failed.Count}"));
            foreach (var r in failed) Ui.Log.Message($"  {Pc.Red("✗")} {r.Skill} → {r.Agent}: {Pc.Dim(r.Error ?? "undefined")}");
        }

        Telemetry.Track(
            ("event", "experimental_sync"),
            ("skillCount", toInstall.Count.ToString()),
            ("successCount", successfulNames.Count.ToString()),
            ("agents", string.Join(",", targetAgents)));

        Sys.OutLine();
        Ui.Outro($"{Pc.Green("Done!")}{Pc.Dim("  Review skills before use; they run with full agent permissions.")}");
    }

    public static SyncOptions ParseOptions(IReadOnlyList<string> args)
    {
        var o = new SyncOptions();
        for (var i = 0; i < args.Count; i++)
        {
            switch (args[i])
            {
                case "-y" or "--yes":
                    o.Yes = true;
                    break;
                case "-f" or "--force":
                    o.Force = true;
                    break;
                case "-a" or "--agent":
                    o.Agent ??= [];
                    while (i + 1 < args.Count && args[i + 1].Length > 0 && !args[i + 1].StartsWith('-')) o.Agent.Add(args[++i]);
                    break;
            }
        }
        return o;
    }
}

File: rust/src/sync.rs
//! `skills experimental_sync` — crawl node_modules for skills (port of sync.ts).

use crate::agents::{
    self, agent, detect_installed_agents, get_non_universal_agents, get_universal_agents,
    get_visible_universal_agents, AgentType,
};
use crate::color::pc;
use crate::detect_agent::{detect_agent, get_agent_type};
use crate::installer::{get_canonical_path, install_skill_for_agent, InstallMode, InstallOptions};
use crate::local_lock::{
    add_skill_to_local_lock, compute_skill_folder_hash, read_local_lock, Entry,
};
use crate::paths::{self, join};
use crate::search_multiselect::{search_multiselect, LockedSection, Options, SearchItem};
use crate::skills::parse_skill_md;
use crate::telemetry::track;
use crate::types::Skill;
use crate::ui::{self, log};
use crate::{outln, sys};
use serde_json::Value;

#[derive(Default, Clone)]
pub struct SyncOptions {
    pub agent: Option<Vec<String>>,
    pub yes: bool,
    pub force: bool,
}

pub fn shorten_path(full: &str, cwd: &str) -> String {
    let home = sys::homedir();
    if full == home || full.starts_with(&format!("{}{}", home, paths::SEP)) {
        return format!("~{}", &full[home.len()..]);
    }
    if full == cwd || full.starts_with(&format!("{}{}", cwd, paths::SEP)) {
        return format!(".{}", &full[cwd.len()..]);
    }
    full.to_string()
}

fn is_dir(p: &str) -> bool {
    std::fs::metadata(p).map(|m| m.is_dir()).unwrap_or(false)
}

fn list_names(dir: &str) -> Option<Vec<String>> {
    Some(
        std::fs::read_dir(dir)
            .ok()?
            .flatten()
            .map(|e| e.file_name().to_string_lossy().to_string())
            .collect(),
    )
}

/// Skills in node_modules packages (top-level and scoped), tagged with the
/// package they came from.
fn discover_node_module_skills(cwd: &str) -> Vec<(Skill, String)> {
    let nm = join(&[cwd, "node_modules"]);
    let mut out = Vec::new();
    let Some(top) = list_names(&nm) else {
        return out;
    };

    let process = |pkg_dir: &str, pkg: &str, out: &mut Vec<(Skill, String)>| {
        let root_md = join(&[pkg_dir, "SKILL.md"]);
        if std::path::Path::new(&root_md).exists() {
            if let Some(s) = parse_skill_md(&root_md, false) {
                out.push((s, pkg.to_string()));
                return;
            }
        }
        for search in [
            pkg_dir.to_string(),
            join(&[pkg_dir, "skills"]),
            join(&[pkg_dir, ".agents", "skills"]),
        ] {
            let Some(names) = list_names(&search) else {
                continue;
            };
            for name in names {
                let skill_dir = join(&[search.as_str(), name.as_str()]);
                if !is_dir(&skill_dir) {
                    continue;
                }
                let md = join(&[skill_dir.as_str(), "SKILL.md"]);
                if !std::path::Path::new(&md).exists() {
                    continue;
                }
                if let Some(s) = parse_skill_md(&md, false) {
                    out.push((s, pkg.to_string()));
                }
            }
        }
    };

    for name in top {
        if name.starts_with('.') {
            continue;
        }
        let full = join(&[nm.as_str(), name.as_str()]);
        if !is_dir(&full) {
            continue;
        }
        if name.starts_with('@') {
            let Some(scoped) = list_names(&full) else {
                continue;
            };
            for sn in scoped {
                let sp = join(&[full.as_str(), sn.as_str()]);
                if is_dir(&sp) {
                    process(&sp, &format!("{}/{}", name, sn), &mut out);
                }
            }
        } else {
            process(&full, &name, &mut out);
        }
    }
    out
}

pub fn run_sync(_args: &[String], mut options: SyncOptions) {
    crate::installer::reset_populated();
    let cwd = sys::cwd();
    let agent_result = detect_agent();
    if agent_result.is_agent() {
        options.yes = true;
        if options.agent.as_ref().map(|a| a.is_empty()).unwrap_or(true) {
            if let Some(mapped) = get_agent_type(agent_result.name()) {
                let mut list: Vec<String> = vec![mapped.to_string()];
                for ua in get_universal_agents() {
                    if !list.iter().any(|x| x == ua) {
                        list.push(ua.to_string());
                    }
                }
                options.agent = Some(list);
            }
        }
    }

    outln!();
    if !agent_result.is_agent() {
        ui::intro(&pc::bg_cyan(pc::black(" skills experimental_sync ")));
    } else {
        log::info(&format!(
            "{} Agent detected — installing non-interactively",
            pc::bg_cyan(pc::black(pc::bold(format!(" {} ", agent_result.name()))))
        ));
    }

    let mut spinner = ui::Spinner::new();
    spinner.start("Scanning node_modules for skills…");
    let discovered = discover_node_module_skills(&cwd);
    if discovered.is_empty() {
        spinner.stop(&pc::yellow("No skills found"));
        ui::outro(&pc::dim("No SKILL.md files found in node_modules."));
        return;
    }
    spinner.stop(&format!(
        "Found {} skill{} in node_modules",
        pc::green(discovered.len().to_string()),
        if discovered.len() > 1 { "s" } else { "" }
    ));

    for (s, pkg) in &discovered {
        log::info(&format!(
            "{} {}",
            pc::cyan(&s.name),
            pc::dim(format!("from {}", pkg))
        ));
        if !s.description.is_empty() {
            log::message(&pc::dim(format!("  {}", s.description)));
        }
    }

    let local_lock = read_local_lock(Some(&cwd));
    let mut to_install: Vec<&(Skill, String)> = Vec::new();
    let mut up_to_date: Vec<String> = Vec::new();
    if options.force {
        to_install.extend(discovered.iter());
        log::info(&pc::dim("Force mode: reinstalling all skills"));
    } else {
        for item in &discovered {
            if let Some(existing) = local_lock.skills.get(&item.0.name) {
                if let Ok(h) = compute_skill_folder_hash(&item.0.path) {
                    if Some(h.as_str()) == existing.get("computedHash").and_then(|v| v.as_str()) {
                        up_to_date.push(item.0.name.clone());
                        continue;
                    }
                }
            }
            to_install.push(item);
        }
        if !up_to_date.is_empty() {
            log::info(&pc::dim(format!(
                "{} skill{} already up to date",
                up_to_date.len(),
                if up_to_date.len() != 1 { "s" } else { "" }
            )));
        }
        if to_install.is_empty() {
            outln!();
            ui::outro(&pc::green("All skills are up to date."));
            return;
        }
    }

    log::info(&format!(
        "{} skill{} to install/update",
        to_install.len(),
        if to_install.len() != 1 { "s" } else { "" }
    ));

    let universal = get_universal_agents();
    let visible_universal = get_visible_universal_agents();
    let locked_section = || LockedSection {
        title: "Universal (.agents/skills)".into(),
        items: visible_universal
            .iter()
            .map(|a| SearchItem::new(*a, a, agent(a).display_name))
            .collect(),
        hidden_count: universal.len() - visible_universal.len(),
    };

    let target_agents: Vec<AgentType> = if options
        .agent
        .as_ref()
        .map(|a| a.iter().any(|x| x == "*"))
        .unwrap_or(false)
    {
        let all = agents::all_agent_names();
        log::info(&format!("Installing to all {} agents", all.len()));
        all
    } else if let Some(list) = options.agent.as_ref().filter(|l| !l.is_empty()) {
        let invalid: Vec<&String> = list.iter().filter(|a| agents::get(a).is_none()).collect();
        if !invalid.is_empty() {
            log::error(&format!(
                "Invalid agents: {}",
                invalid
                    .iter()
                    .map(|s| s.as_str())
                    .collect::<Vec<_>>()
                    .join(", ")
            ));
            log::info(&format!(
                "Valid agents: {}",
                agents::all_agent_names().join(", ")
            ));
            sys::exit(1);
        }
        list.iter()
            .map(|a| agents::to_agent_type(a).unwrap())
            .collect()
    } else {
        spinner.start("Loading agents…");
        let installed = detect_installed_agents();
        spinner.stop(&format!("{} agents", agents::agents().len()));
        if installed.is_empty() {
            if options.yes {
                log::info("Installing to universal agents");
                universal.clone()
            } else {
                let items: Vec<SearchItem<AgentType>> = get_non_universal_agents()
                    .into_iter()
                    .map(|a| {
                        SearchItem::new(a, a, agent(a).display_name)
                            .hint(Some(agent(a).skills_dir.to_string()))
                    })
                    .collect();
                let mut o = Options::new("Which agents do you want to install to?", items);
                o.locked_section = Some(locked_section());
                match search_multiselect(o) {
                    Some(v) => v,
                    None => {
                        ui::cancel("Sync cancelled");
                        sys::exit(0);
                    }
                }
            }
        } else if installed.len() == 1 || options.yes {
            let mut t = installed.clone();
            for ua in &universal {
                if !t.contains(ua) {
                    t.push(ua);
                }
            }
            t
        } else {
            let others: Vec<AgentType> = get_non_universal_agents()
                .into_iter()
                .filter(|a| installed.contains(a))
                .collect();
            let items: Vec<SearchItem<AgentType>> = others
                .iter()
                .map(|a| {
                    SearchItem::new(*a, a, agent(a).display_name)
                        .hint(Some(agent(a).skills_dir.to_string()))
                })
                .collect();
            let initial: Vec<usize> = installed
                .iter()
                .filter(|a| !universal.contains(a))
                .filter_map(|a| others.iter().position(|o| o == a))
                .collect();
            let mut o = Options::new("Which agents do you want to install to?", items);
            o.initial_selected = initial;
            o.locked_section = Some(locked_section());
            match search_multiselect(o) {
                Some(v) => v,
                None => {
                    ui::cancel("Sync cancelled");
                    sys::exit(0);
                }
            }
        }
    };

    let mut summary: Vec<String> = Vec::new();
    for (s, pkg) in &to_install {
        let canonical = get_canonical_path(&s.name, false, None, None, None).unwrap_or_default();
        summary.push(format!(
            "{} {}",
            pc::cyan(&s.name),
            pc::dim(format!("← {}", pkg))
        ));
        summary.push(format!("  {}", pc::dim(shorten_path(&canonical, &cwd))));
    }
    outln!();
    ui::note(&summary.join("\n"), "Sync Summary");

    if !options.yes {
        match ui::confirm("Proceed with sync?", true) {
            Some(true) => {}
            _ => {
                ui::cancel("Sync cancelled");
                sys::exit(0);
            }
        }
    }

    spinner.start("Syncing skills…");
    struct R {
        skill: String,
        agent: String,
        success: bool,
        canonical: Option<String>,
        error: Option<String>,
    }
    let mut results: Vec<R> = Vec::new();
    for (s, _) in &to_install {
        for at in &target_agents {
            let r = install_skill_for_agent(
                s,
                at,
                &InstallOptions {
                    global: false,
                    cwd: Some(cwd.clone()),
                    mode: Some(InstallMode::Symlink),
                    ..Default::default()
                },
            );
            results.push(R {
                skill: s.name.clone(),
                agent: agent(at).display_name.to_string(),
                success: r.success,
                canonical: r.canonical_path,
                error: r.error,
            });
        }
    }
    spinner.stop("Sync complete");

    let successful_names: Vec<String> = {
        let mut v: Vec<String> = Vec::new();
        for r in results.iter().filter(|r| r.success) {
            if !v.contains(&r.skill) {
                v.push(r.skill.clone());
            }
        }
        v
    };

    for (s, pkg) in &to_install {
        if successful_names.contains(&s.name) {
            if let Ok(h) = compute_skill_folder_hash(&s.path) {
                let mut e = Entry::new();
                e.insert("source".into(), Value::String(pkg.clone()));
                e.insert("sourceType".into(), Value::String("node_modules".into()));
                e.insert("computedHash".into(), Value::String(h));
                let _ = add_skill_to_local_lock(&s.name, e, Some(&cwd));
            }
        }
    }

    outln!();
    if !successful_names.is_empty() {
        let mut lines = Vec::new();
        for name in &successful_names {
            let first = results
                .iter()
                .find(|r| r.success && &r.skill == name)
                .unwrap();
            let pkg = to_install
                .iter()
                .find(|(s, _)| &s.name == name)
                .map(|(_, p)| p.clone())
                .unwrap_or_default();
            lines.push(format!(
                "{} {} {}",
                pc::green("✓"),
                name,
                pc::dim(format!("← {}", pkg))
            ));
            if let Some(c) = &first.canonical {
                lines.push(format!("  {}", pc::dim(shorten_path(c, &cwd))));
            }
        }
        let n = successful_names.len();
        ui::note(
            &lines.join("\n"),
            &pc::green(format!(
                "Synced {} skill{}",
                n,
                if n != 1 { "s" } else { "" }
            )),
        );
    }
    let failed: Vec<&R> = results.iter().filter(|r| !r.success).collect();
    if !failed.is_empty() {
        outln!();
        log::error(&pc::red(format!("Failed to install {}", failed.len())));
        for r in failed {
            log::message(&format!(
                "  {} {} → {}: {}",
                pc::red("✗"),
                r.skill,
                r.agent,
                pc::dim(r.error.clone().unwrap_or_else(|| "undefined".into()))
            ));
        }
    }

    track(&[
        ("event", Some("experimental_sync".into())),
        ("skillCount", Some(to_install.len().to_string())),
        ("successCount", Some(successful_names.len().to_string())),
        ("agents", Some(target_agents.join(","))),
    ]);

    outln!();
    ui::outro(&format!(
        "{}{}",
        pc::green("Done!"),
        pc::dim("  Review skills before use; they run with full agent permissions.")
    ));
}

pub fn parse_sync_options(args: &[String]) -> SyncOptions {
    let mut o = SyncOptions::default();
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "-y" | "--yes" => o.yes = true,
            "-f" | "--force" => o.force = true,
            "-a" | "--agent" => {
                let list = o.agent.get_or_insert_with(Vec::new);
                while i + 1 < args.len() && !args[i + 1].is_empty() && !args[i + 1].starts_with('-')
                {
                    i += 1;
                    list.push(args[i].clone());
                }
            }
            _ => {}
        }
        i += 1;
    }
    o
}
